addons/udes_stock/models/stock_picking.py

```diff
@@ -1074,3 +1091,33 @@ def unlink(self):
                     % ", ".join(non_draft_picks.mapped("name"))
                 )
         return super().unlink()
+
+    def write(self, vals):
+        """
+        If odoo attempts to add move line already on picking, remove the action from vals.
+        When saving a picking from the desktop odoo will attempt to re-add all move_lines
+        to the picking even when they already exist on the picking. This causes issues
+        when an order has some move lines completed (with real_time flag = True) and some
+        needing completion. In UDES we will assign any move lines needing completion
+        to the backorder picking but core Odoo will then reassign these move lines to the
+        original picking (due to the extra actions in vals).
+        """
+
+        # Only consider removing commands when validating in real time and picking is in assigned state
+        pickings_to_check = self.filtered(
+            lambda p: p.picking_type_id.u_validate_real_time and self.state == "assigned"
+        )
+        for picking in pickings_to_check:
+            move_lines_on_pick = set(picking.move_line_ids.ids)
+            actions = vals.get("move_line_ids")
+            if actions:
+                new_actions = list()
+                for action in actions:
+                    command = action[0]
+                    move_line_id = action[1]
+                    # 4 is command for appending one record to another recordset
+                    if not (command == 4 and move_line_id in move_lines_on_pick):
```

Comment:
Maybe it looks easier to read in this way
```suggestion
                    if command != 4 and move_line_id not in move_lines_on_pick:
```

Reply:
changed https://github.com/unipartdigital/udes-open/pull/801/commits/042397f7cd7a41306c6e6373258a44d16272d0c7

Reply:
Sorry just checked and this is wrong, because `not (command == 4 and move_line_id in move_lines_on_pick) ` is not the same logic as `command != 4 and move_line_id not in move_lines_on_pick`

Reply:
I am going to drop this commit

Reply:
![image](https://github.com/unipartdigital/udes-open/assets/95625678/e83b9701-ef7e-4c70-9f32-4e83655ea571)